src/Compiler/Microsoft.AspNetCore.Razor.Language/src/MetadataCollection.cs

```diff
@@ -79,85 +82,63 @@ public static MetadataCollection Create(IReadOnlyList<KeyValuePair<string, strin
-                    {
-                        throw new InvalidOperationException();
-                    }
-
-                    var pair = enumerable.Current;
-
-                    return new OneToThreeItems(pair.Key, pair.Value);
-                }
-
-            case 2:
-                {
-                    using var enumerable = map.GetEnumerator();
-
-                    if (!enumerable.MoveNext())
-                    {
-                        throw new InvalidOperationException();
-                    }
-
-                    var pair1 = enumerable.Current;
-
-
-                    if (!enumerable.MoveNext())
-                    {
-                        throw new InvalidOperationException();
-                    }
-
-                    var pair2 = enumerable.Current;
-
-                    return new OneToThreeItems(pair1.Key, pair1.Value, pair2.Key, pair2.Value);
-                }
+            [] => Empty,
+            [var pair] => new OneToThreeItems(pair.Key, pair.Value),
+            [var pair1, var pair2] => new OneToThreeItems(pair1.Key, pair1.Value, pair2.Key, pair2.Value),
+            [var pair1, var pair2, var pair3] => new OneToThreeItems(pair1.Key, pair1.Value, pair2.Key, pair2.Value, pair3.Key, pair3.Value),
+            _ => new FourOrMoreItems(pairs),
+        };
 
-            case 3:
-                {
-                    using var enumerable = map.GetEnumerator();
+    public static MetadataCollection Create(IReadOnlyList<KeyValuePair<string, string?>> pairs)
+        => pairs switch
+        {
+            [] => Empty,
+            [var pair] => new OneToThreeItems(pair.Key, pair.Value),
+            [var pair1, var pair2] => new OneToThreeItems(pair1.Key, pair1.Value, pair2.Key, pair2.Value),
+            [var pair1, var pair2, var pair3] => new OneToThreeItems(pair1.Key, pair1.Value, pair2.Key, pair2.Value, pair3.Key, pair3.Value),
+            _ => new FourOrMoreItems(pairs),
+        };
 
-                    if (!enumerable.MoveNext())
-                    {
-                        throw new InvalidOperationException();
-                    }
+    public static MetadataCollection Create(Dictionary<string, string?> map)
+    {
+        using var builder = new PooledArrayBuilder<KeyValuePair<string, string?>>(capacity: map.Count);
 
-                    var pair1 = enumerable.Current;
+        foreach (var pair in map)
+        {
+            builder.Add(pair);
+        }
 
-                    if (!enumerable.MoveNext())
-                    {
-                        throw new InvalidOperationException();
-                    }
+        return Create(builder.DrainToImmutable());
+    }
 
-                    var pair2 = enumerable.Current;
+    public static MetadataCollection Create(IReadOnlyDictionary<string, string?> map)
+    {
+        // Take a faster path if Dictionary<string, string?> is passed to us. This ensures that
+        // we use Dictionary's struct-based enumerator.
```

Comment:
> struct-based enumerator

The optimization seems to be larger than that - in the other case we don't use *any* enumerator. Should we though instead of calling `ToArray()`?

Reply:
`ToArray()` will ultimately use the enumerator of any `IReadOnlyDictionary<>`, no?